### PR TITLE
Python Interface for Jiterator

### PR DESCRIPTION
**Pull request by SherlockNoMad**
This PR allows user to author a CUDA kernel in python. 

```
from torch.cuda.jiterator import create_jit_fn

code_string = "template <typename T> T my_kernel(T x, T y, T alpha) { return  -x * y + x - y + alpha; }"
jitted_fn = create_jit_fn(code_string, alpha=0)

a = torch.rand(3, device='cuda')
b = torch.rand(3, device='cuda')
result = jitted_fn(a, b, alpha=1.0)
```

Limitations: 
- Only supports elementwise kernel
- 1~8 tensor inputs (empty input, e.g. factory methods, is not supported)
- inputs tensors must live in cuda device
- cpu Scalar is not supported
- kwargs must be pre-declared when calling create_jit_fn
- kwargs must be convertible to at::Scalar, one of float64, int64_t, bool. (complex not support for now)

TODOs: 
- [x] consolidate union and c10::variant implementation
- [x] plug into existing op testing framework
- [ ] rename files, place files in the right folder
- [ ] place util functions in the right file
- [x] enforce assumptions in python interface e.g <8 inputs, kwargs types
- [x] Add user-facing documentation

### CODE REVIEW

File: test/test_jiterator.py
Comment (SherlockNoMad) [inline test/test_jiterator.py:34]: already covered in OpInfo, can remove?

File: test/test_jiterator.py
Comment (SherlockNoMad) [inline test/test_jiterator.py:65]: covered in opinfo, can remove?

File: torch/_C/__init__.pyi.in
Comment (SherlockNoMad) [inline torch/_C/__init__.pyi.in:911]: bool type kwargs is not tested yet...

File: test/test_jiterator.py
Comment (SherlockNoMad) [inline test/test_jiterator.py:96]: covered in opinfo?

File: aten/src/ATen/cuda/jiterator.cu
Comment (SherlockNoMad) [inline aten/src/ATen/cuda/jiterator.cu:6]: @ngimel, @mruberry 
Any suggestion on the file names and file dir placements?

File: aten/src/ATen/cuda/jiterator.cu
Comment (SherlockNoMad) [inline aten/src/ATen/cuda/jiterator.cu:190]: remove.

File: aten/src/ATen/cuda/jiterator.cu
Comment (SherlockNoMad) [inline aten/src/ATen/cuda/jiterator.cu:330]: @ngimel, @mruberry
Not sure if it makes a difference in this case?

File: aten/src/ATen/native/cuda/jit_utils.h
Comment (SherlockNoMad) [inline aten/src/ATen/native/cuda/jit_utils.h:110]: Why?

Comment (ngimel) [reply aten/src/ATen/native/cuda/jit_utils.h:110]: No particular reason, we didn't have jiterated functions that had scalar_t bool?

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2283]: Cool!

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:2300]: fyi, a common pattern which might save you some typing in the future is to partial `make_tensor` to always insert the device, dtype, and requires_grad info

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:15749]: prefer expectedFailure to skip

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:15739]: Add a comment explaining what these OpInfos are for and pointing people to the jiterator note

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:15778]: Testing the jiterator by modeling these as ops is a neat idea

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:15756]: Nice idea to have one sample op per class of operation the jiterator can generate

File: torch/cuda/jiterator.py
Comment (mruberry) [inline torch/cuda/jiterator.py:43]: Is there something we should refer users to to understand the jiterator? Or does this need to elaborate on what the jiterator and its properties are?

File: torch/cuda/jiterator.py
Comment (mruberry) [inline torch/cuda/jiterator.py:58]: Nice warning.

File: torch/cuda/jiterator.py
Comment (mruberry) [inline torch/cuda/jiterator.py:50]: This example is very good. Additional details about the form the CUDA string must take would be helpful. Like, it's a templated function that accepts ...

File: torch/cuda/jiterator.py
Comment (mruberry) [inline torch/cuda/jiterator.py:19]: Is all this parsing necessary?

File: torch/cuda/jiterator.py
Comment (mruberry) [reply torch/cuda/jiterator.py:43]: Follow-up question: should we also direct users to the cache? In the future, do you think we should add more cache controls, like giving users the ability to specify the cache location in Python or clear it from Python? Triton and JAX have some additional cache capabilities

File: torch/csrc/cuda/Module.cpp
Comment (mruberry) [inline torch/csrc/cuda/Module.cpp:221]: Add a comment explaining what this function is for

File: torch/csrc/cuda/Module.cpp
Comment (mruberry) [inline torch/csrc/cuda/Module.cpp:242]: add a comment

File: tools/linter/adapters/newlines_linter.py
Comment (mruberry) [inline tools/linter/adapters/newlines_linter.py:70]: Nice spelling fixes

File: test/test_jiterator.py
Comment (mruberry) [inline test/test_jiterator.py:9]: This can be imported from common_utils:

https://github.com/pytorch/pytorch/blob/201ddafc22e22c387b4cd654f397e05354d73d09/torch/testing/_internal/common_utils.py#L763

File: test/test_jiterator.py
Comment (mruberry) [inline test/test_jiterator.py:4]: Does this need to be guarded on CUDA being available?

File: test/test_jiterator.py
Comment (mruberry) [inline test/test_jiterator.py:44]: Use self.assertEqual (or, if you want more control, torch.testing.assert_close)

File: test/test_jiterator.py
Comment (mruberry) [reply test/test_jiterator.py:65]: Yes, this should be redundant with the generated reference tests

File: test/test_jiterator.py
Comment (mruberry) [reply test/test_jiterator.py:96]: Only for elementwise binary OpInfos

**Review by mruberry — COMMENTED** [review]
Hey @SherlockNoMad! This looks really cool; I made a few comments for your review.

@ngimel suggested the three of us might be most productive reviewing it together, so I'll hold off on more for now and see you offline!

File: aten/src/ATen/cuda/jiterator.cu
Comment (SherlockNoMad) [inline aten/src/ATen/cuda/jiterator.cu:276]: comment: copied from cuda loop, why

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:15763]: TODO: delete me?

File: torch/cuda/jiterator.py
Comment (SherlockNoMad) [inline torch/cuda/jiterator.py:4]: should fail if cuda does not exist

**Review by mruberry — APPROVED** [review]
A few notes:
- prototype
- more comments
- self.assertEqual in tests
- update docs to explain the exact type of string that must be specified and what is currently supported more precisely

File: test/test_jiterator.py
Comment (SherlockNoMad) [reply test/test_jiterator.py:9]: right.

File: torch/testing/_internal/common_methods_invocations.py
Comment (SherlockNoMad) [reply torch/testing/_internal/common_methods_invocations.py:2300]: Thanks!

File: torch/testing/_internal/common_methods_invocations.py
Comment (SherlockNoMad) [reply torch/testing/_internal/common_methods_invocations.py:15749]: Update to expectedFailure.

File: test/test_jiterator.py
Comment (SherlockNoMad) [reply test/test_jiterator.py:44]: ok.

File: test/test_jiterator.py
Comment (SherlockNoMad) [reply test/test_jiterator.py:4]: Checked below.

File: test/test_jiterator.py
Comment (SherlockNoMad) [reply test/test_jiterator.py:34]: removed.

File: test/test_jiterator.py
Comment (SherlockNoMad) [reply test/test_jiterator.py:96]: updated.

File: test/test_jiterator.py
Comment (SherlockNoMad) [reply test/test_jiterator.py:44]: fixed.

File: torch/cuda/jiterator.py
Comment (SherlockNoMad) [reply torch/cuda/jiterator.py:4]: following cuda module's lazy initialization pattern. 
checked in __init__ function.

File: torch/csrc/cuda/Module.cpp
Comment (SherlockNoMad) [reply torch/csrc/cuda/Module.cpp:221]: done.

File: torch/testing/_internal/common_methods_invocations.py
Comment (SherlockNoMad) [reply torch/testing/_internal/common_methods_invocations.py:15739]: done.

File: torch/cuda/jiterator.py
Comment (SherlockNoMad) [reply torch/cuda/jiterator.py:50]: Done.

File: torch/_C/__init__.pyi.in
Comment (SherlockNoMad) [reply torch/_C/__init__.pyi.in:911]: checked, it's working.

File: aten/src/ATen/cuda/jiterator.cu
Comment (SherlockNoMad) [reply aten/src/ATen/cuda/jiterator.cu:276]: done.

**Review by suo — CHANGES_REQUESTED** [review]
sorry for the drive by, but I noticed while messing with the stats that the test file added by this PR adds almost 1h on average to our test suite!!! [query-results (7).csv](https://github.com/pytorch/pytorch/files/8618739/query-results.7.csv). The earlier runs were faster, but the later ones are slower, I guess because you added more tests maybe?

That is enough to be land blocking I think; do you mind taking a look before merging?

**Comment by SherlockNoMad** [comment]
> sorry for the drive by, but I noticed while messing with the stats that the test file added by this PR adds almost 1h on average to our test suite!!! [query-results (7).csv](https://github.com/pytorch/pytorch/files/8618739/query-results.7.csv). The earlier runs were faster, but the later ones are slower, I guess because you added more tests maybe?
> 
> That is enough to be land blocking I think; do you mind taking a look before merging?

Thanks @suo for catching this surprising issue!! The entire test takes ~10s on my local dev box.  

Following are the test cases that took longer than 10s. 
```
  test_all_dtype_and_layout_a_shape_stride_([3, 1], [1, 3])_b_shape_stride_([3, 3], [3, 1])_cuda_bfloat16_complex128 (__main__.TestPythonJiteratorCUDA) ... ok (361.322s)
  test_all_dtype_and_layout_a_shape_stride_([3, 1], [1, 3])_b_shape_stride_([3, 3], [3, 1])_cuda_bfloat16_complex64 (__main__.TestPythonJiteratorCUDA) ... ok (676.216s)
  test_all_dtype_and_layout_a_shape_stride_([3, 1], [1, 3])_b_shape_stride_([3, 3], [3, 1])_cuda_bfloat16_float16 (__main__.TestPythonJiteratorCUDA) ... ok (288.199s)
  test_all_dtype_and_layout_a_shape_stride_([3, 1], [1, 3])_b_shape_stride_([3, 3], [3, 1])_cuda_bfloat16_float64 (__main__.TestPythonJiteratorCUDA) ... ok (287.869s)
  test_all_dtype_and_layout_a_shape_stride_([3, 1], [1, 3])_b_shape_stride_([3, 3], [3, 1])_cuda_bfloat16_int16 (__main__.TestPythonJiteratorCUDA) ... ok (296.899s)
  test_all_dtype_and_layout_a_shape_stride_([3, 1], [1, 3])_b_shape_stride_([3, 3], [3, 1])_cuda_float16_int16 (__main__.TestPythonJiteratorCUDA) ... ok (287.556s)
  test_all_dtype_and_layout_a_shape_stride_([3, 1], [1, 3])_b_shape_stride_([3, 3], [3, 1])_cuda_int16_int32 (__main__.TestPythonJiteratorCUDA) ... ok (290.443s)
  test_all_dtype_and_layout_a_shape_stride_([3, 1], [1, 3])_b_shape_stride_([3, 3], [3, 1])_cuda_int16_int64 (__main__.TestPythonJiteratorCUDA) ... ok (286.148s)
  test_all_dtype_and_layout_a_shape_stride_([3, 1], [1, 3])_b_shape_stride_([3, 3], [3, 1])_cuda_int16_int8 (__main__.TestPythonJiteratorCUDA) ... ok (287.225s)
```

As jiterator compiles cuda kernel on the fly, I suspect these slow tests are caused by the nvrtc compiler. I will need to ssh into the CI machine for a repro. 

@ngimel, @mruberry All thees long running test cases have "noncontiguous" inputs and requires "dynamic casting". Have we encounter similar issues before?

**Comment by ngimel** [comment]
This is unexpected! Yes, noncontiguous strides and dynamic casting takes longer to compile, but it should be on the order of seconds, not 300 seconds! @SherlockNoMad is it possible that on your devbox you are benefitting from cached kernels that CI doesn't have? That could explain the difference. But, again, 300s time is totally unexpected.

**Comment by SherlockNoMad** [comment]
I have confirmed that nvrtcCompileProgram call is taking really long in some cases...

In the following case, it's taking 271s to compile!!! 

Generated kernel: https://gist.github.com/SherlockNoMad/7bcf71939de0c4bd7dc4c023c04433e1 

```
2022-05-04T20:32:26.1895195Z   test_all_dtype_and_layout_shape_strides_(([3, 3], [1, 3]), ([3, 1], [1, 3]))_cuda_bfloat16_float16 (__main__.TestPythonJiteratorCUDA) ... 
2022-05-04T20:32:26.1912726Z nvrtcCompileProgram begins
2022-05-04T20:36:58.1158786Z nvrtcCompileProgram time = 271924442[µs]
2022-05-04T20:36:58.1364562Z ok (271.949s)
```

**Comment by ptrblck** [comment]
Yes, we should definitely follow up on this issue internally.
I'm trying to reproduce the long compile time by building your branch and running:
```python
time python test_jiterator.py -v
```
which gives:
```python
Ran 168 tests in 21.540s

OK

real    0m25.836s
user    0m22.591s
sys     0m10.504s
```
I assume you might be using an older CUDA toolkit, which might show the issue or I'm missing something in the reproduction.
Could you let me know which CUDA release you are using, please?

**Comment by SherlockNoMad** [comment]
With Cuda11.6, nvrtcCompileProgram doesn't seem to have this problem. 

```
test_all_dtype_and_layout_shape_strides_(([3, 3], [1, 3]), ([3, 1], [1, 3]))_cuda_bfloat16_float16 (__main__.TestPythonJiteratorCUDA) ... 
nvrtcCompileProgram begins
nvrtcCompileProgram time = 3932783[µs]
ok (3.939s)
```

The entire test passed in 60 second. 

```
Ran 168 tests in 52.402s
```

See https://github.com/pytorch/pytorch/runs/6311742745?check_suite_focus=true for details. 


@suo, @ngimel, I am going to limit this test case to be only run on CUDA >= 11.6 as a workaround.

**Comment by SherlockNoMad** [comment]
@ptrblck, @ngimel 
Just to mention another data point. 

On my dev machine, it's using CUDA11.1, and nvrtcCompileProgram is not having long running problem.  
I am wondering if it's a problem specific to CUDA11.3, or there is something special about the CI build env.

**Comment by SherlockNoMad** [comment]
With the workaround, test_jiterators completes in reasonable time. 
```
Ran 289 tests in 45.547s
```

Closing the PR now...

**Pull request closed without merging**
